src/registry/asynchronous.rs

```diff
@@ -1,46 +1,140 @@
-use futures::stream::Stream;
-use futures::stream::{self, StreamExt};
+use std::{iter::Rev, ops::Range};
+
+use super::{L10nRegistry, L10nRegistryLocked};
+use crate::{fluent::FluentBundle, futures::AllSome, source::RcResource, source::ResourceStatus};
+
+use futures::{FutureExt, Stream, ready};
 use unic_langid::LanguageIdentifier;
 
-use super::L10nRegistry;
-use crate::fluent::FluentBundle;
+pub type ResourceSetStream = AllSome<ResourceStatus, RcResource>;
 
-impl L10nRegistry {
-    pub fn generate_bundles_for_lang<'l, P>(
-        &'l self,
-        langid: &'l LanguageIdentifier,
-        res_ids: impl IntoIterator<Item = P> + Clone + 'l,
-    ) -> impl Stream<Item = FluentBundle> + 'l
+impl<'a> L10nRegistryLocked<'a> {
+    pub(crate) fn generate_resource_set<P>(
+        &self,
+        langid: &LanguageIdentifier,
+        source_order: &[usize],
+        resource_ids: &[P],
+    ) -> ResourceSetStream
     where
-        P: AsRef<str> + Clone + 'l,
+        P: AsRef<str>,
     {
-        let permutations = self
-            .generate_source_permutations(langid, res_ids.clone())
-            .map(move |sources| sources.into_iter().zip(res_ids.clone()));
-
-        stream::iter(permutations).filter_map(move |sources| async move {
-            let mut bundle = FluentBundle::new(&[langid.clone()]);
-            for (source, res_id) in sources {
-                if let Some(res) = source.fetch_file(&langid, res_id.as_ref()).await {
-                    bundle.add_resource(res).unwrap();
-                } else {
-                    return None;
+        debug_assert_eq!(source_order.len(), resource_ids.len());
+        source_order
+            .iter()
+            .zip(resource_ids.iter().map(AsRef::as_ref))
+            .map(|(&idx, path)| self.source_idx(idx).fetch_file(langid, path))
+            .collect()
+    }
+}
+
+impl L10nRegistry {
+    pub fn generate_bundles_for_lang(
+        &self,
+        langid: LanguageIdentifier,
+        resource_ids: Vec<String>,
+    ) -> GenerateBundles {
+        let lang_ids = vec![langid];
+
+        GenerateBundles::new(self.clone(), lang_ids, resource_ids)
+    }
+
+    pub fn generate_bundles(
+        &self,
+        lang_ids: Vec<LanguageIdentifier>,
+        resource_ids: Vec<String>,
+    ) -> GenerateBundles {
+        let lang_ids = lang_ids.into_iter().collect();
+        let resource_ids = resource_ids.into_iter().collect();
```

Comment:
why do you do this here? Do you need to clone it?

Reply:
Left over code from a previous changes. Removed.

Reply:
That code isn't needed anymore. Removed.